### ISSUE TITLE
Unreadable Content

### ISSUE THREAD

**Issue by jwerre** [Ziv-Barber/officegen#72]
After generating a Word Doc I get the following errors when trying to open it.

![screen shot 2015-07-24 at 11 37 28 am](https://cloud.githubusercontent.com/assets/747587/8882461/e2b9c0d4-31fb-11e5-96dd-d985e68a0055.png)

![screen shot 2015-07-24 at 11 38 07 am](https://cloud.githubusercontent.com/assets/747587/8882462/e2ccdbce-31fb-11e5-8a38-a90f5542fbbf.png)

This is a very simple document with a single word  in it. Do you have any idea what could be causing this?

**Comment by pietersv** [comment]
Quick checks, do the following work if you clone the repo and run these from the project root?

```
cd ~/MyProjects/
git clone https://github.com/Ziv-Barber/officegen.git
cd officegen
npm test // should return all passing

node example/make_docx.js
open out.docx // should open successfully
```
Can you post your code that generates the doc?

**Comment by jwerre** [comment]
Ah... I set the 'a' flag on my write stream which was causing the error.

   var exportStream = fs.createWriteStream(@exportPath, {flags: 'a'})

**Comment by jwerre** [comment]
I am actually still having this issue but I don't think it has anything directly to do with officegen.

If I generate the file directly with officegen I am able to open it:

	var filename = 'my_doc.docx',
		ws = fs.createWriteStream(filename),
		docx = officegen({type: 'docx', onend:done}),
		p = docx.createP()
	p.addText( data.survey.title, @styles.h1 )
	docx.generate( ws )

The document generated here seems to work fine. However if I put it in an express route and download it via the web browser I get the above error.

	router.get( "export/docx", function(req, res, next){
		res.download( filename )
	})

The file seems to have the appropriate header information so I'm not sure what could be causing the problem:

	HTTP/1.1 200 OK
	X-Powered-By: Express
	Content-Disposition: attachment; filename="my_doc.docx"
	Accept-Ranges: bytes
	Cache-Control: public, max-age=0
	Last-Modified: Thu, 30 Jul 2015 20:17:44 GMT
	ETag: W/"201d-14ee09d0640"
	Content-Type: application/vnd.openxmlformats-officedocument.wordprocessingml.document
	Content-Length: 8221
	Date: Thu, 30 Jul 2015 20:17:44 GMT
	Connection: keep-alive

**Comment by protobi** [comment]
Try using `docx.generate(res)` 
```js
   send: function (res, options) {

      res.set({
        "Content-Type": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        'Content-disposition': 'attachment; filename=' + this.filename
      });
      docx.generate(res);
    }
```

**Comment by jwerre** [comment]
I thought this might be the issue but I *do* have these headers set. See the header information above.

**Comment by protobi** [comment]
The error message includes a backslash, might this be an issue with Windows vs OSX/Linux?  Did you try having officegen stream the doc directly to the express `res` object vs streaming the file to disk and downloading it?

**Comment by jwerre** [comment]
Yes, streaming it directly seems to to fix it. But I'm sure I don't have any backslashes in the the file name. What do you mean by "The error message includes a backslash"?

**Comment by pietersv** [comment]
Glad that worked!  Was just grasping at straws, based on this phrase below.  

> "...the filename might contain invalid characters (for example, <>|\"/)"

Have in past in other (non-office) projects run into issues with files not transferring nicely between *nix and Windows.  I wish Office provided more helpful diagnostic messages.

**Comment by pietersv** [comment]
It seems you can generate a great matched tria of files for debugging:
* One docx saved to disk on the file server, which opens fine on your machine.
* That same file, sent by Express via `res.download(file)` and saved by client, which doesn't open.
* That same docx, sent by Express via `docx.generate(res)` and saved by client, which does open.

Can you generate such a trio for debugging, minus any confidential info, and send directly? This is outside my domain, but I wrote a small a tool for differencing and isolating problems in Office files.

**Comment by jwerre** [comment]
Ya, I'll do that tomorrow and post it. Mocha tests be ok?

**Comment by Salindakrish** [comment]
@jwerre  did you manage to stream the file to the front end, I am having the same issue. If you succeeded please let me know...

**Comment by edencorbin** [comment]
This post helped me quite a bit, and I saw the recent question @Salindakrish, so I figured I would post the code that worked for me, perhaps its helpful:

```
app.get('/test', function (req, res){
    var docx = officegen ({
        'type': 'docx',
        'subject': 'testing it',
        'keywords': 'some keywords',
        'description': 'a description'
    });

    var pObj = docx.createP ();
    pObj.addText ( 'a very simple paragraph' );

    res.set({
        "Content-Type": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        'Content-disposition': 'attachment; filename=blah.docx'
    });
    docx.generate(res);
});
```

**Comment by Salindakrish** [comment]
thank you @edencorbin .. I did the same but how did you manage to retrieve it from the front end ,,, I may be having a little knowledge  on it,, i used a Blob but it didn't work for me,,,,

`post("/generateppt", {req: '', res: ''})
            .then(function(response) {
                    var pdfDoc = response.data;
               var pdfBlob = new Blob([pdfDoc], { type: "application/vnd.openxmlformats-officedocument.presentationml.presentation" });
                 var pdfURL = window.URL.createObjectURL(pdfBlob);
                var tempLink = document.createElement("a");
                 tempLink.href = pdfURL;
                 tempLink.setAttribute(
                    "download",
                    "Summary"
                 );
                 tempLink.setAttribute("target", "_blank");
                 document.body.appendChild(tempLink);
                 tempLink.click();
                 document.body.removeChild(tempLink);
               
            });`


But this gives a file but ,, after downloading it.. It shows a corrupted file... can you please assist me... I used front end javascript. I created a pptx...

**Comment by edencorbin** [comment]
@Salindakrish I'm not super knowledgeable either on officegen or the method you are attempting to pull it, so cannot help with your error. For me, the standard res from express worked, as shown in my code, although I have not tested it in many browsers, and likely will need to, to ensure compatibility. When ressing from express everything is set on the node/server side, and literally the browser just needs to hit the api url, no code on the client side.

**Comment by Salindakrish** [comment]
@edencorbin  Thanx mate. So did you bind that URL to a button click..??

**Comment by ipopa** [comment]
@Salindakrish

angular example:

          $http({
                    method: "GET",
                    url: 'url/test',
                    responseType: 'arraybuffer'
                })
                .then(function (rs) {
                    saveAs(new Blob([rs.data], {type: "application/vnd.openxmlformats-officedocument.wordprocessingml.document"}), 'test.docx');
                });
